mountpoint-s3/tests/fuse_tests/write_test.rs

```diff
@@ -732,41 +766,151 @@ where
+#[test_case(true; "write_only")]
+#[test_case(false; "read_write")]
+fn overwrite_fail_on_write_without_truncate_test_s3(write_only: bool) {
+    overwrite_fail_on_write_without_truncate_test(
+        fuse::s3_session::new,
+        "overwrite_fail_on_write_without_truncate_test",
+        write_only,
+    );
+}
+
+#[test_case(true; "write_only")]
+#[test_case(false; "read_write")]
+fn overwrite_fail_on_write_without_truncate_test_mock(write_only: bool) {
+    overwrite_fail_on_write_without_truncate_test(
+        fuse::mock_session::new,
+        "overwrite_fail_on_write_without_truncate_test",
+        write_only,
+    );
+}
 
-    // Open with O_RDWR and write something to the file
+fn overwrite_no_truncate_test<F>(creator_fn: F, prefix: &str, write_only: bool)
+where
+    F: FnOnce(&str, TestSessionConfig) -> (TempDir, BackgroundSession, TestClientBox),
+{
+    let filesystem_config = S3FilesystemConfig {
+        allow_overwrite: true,
+        ..Default::default()
+    };
+    let test_config = TestSessionConfig {
+        filesystem_config,
+        ..Default::default()
+    };
+    let (mount_point, _session, mut test_client) = creator_fn(prefix, test_config);
+
+    // Make sure there's an existing directory and a file
+    test_client.put_object("dir/hello.txt", b"hello world").unwrap();
+
+    let _subdir = mount_point.path().join("dir");
+    let path = mount_point.path().join("dir/hello.txt");
+
+    // Open the file in non-truncate mode and do nothing
     let mut options = File::options();
-    let mut write_fh = options.read(true).write(true).truncate(true).open(&path).unwrap();
-    write_fh.write_all(b"first overwrite").expect("write should succeed");
-    write_fh.sync_all().unwrap();
+    if !write_only {
+        options.read(true);
+    }
+    let write_fh = options.write(true).open(&path).expect("open should succeed");
     drop(write_fh);
 
-    // Check the new file content
+    // File content should not be changed
     let mut options = File::options();
-    let mut read_fh = options.read(true).open(&path).unwrap();
+    let mut fh = options.read(true).write(true).open(&path).unwrap();
     let mut hello_contents = String::new();
-    read_fh.read_to_string(&mut hello_contents).unwrap();
-    assert_eq!(hello_contents, "first overwrite");
-    drop(read_fh);
+    fh.read_to_string(&mut hello_contents).unwrap();
+    assert_eq!(hello_contents, "hello world");
+}
+
+#[cfg(feature = "s3_tests")]
+#[test_case(true; "write_only")]
+#[test_case(false; "read_write")]
+fn overwrite_no_truncate_test_s3(write_only: bool) {
+    overwrite_no_truncate_test(fuse::s3_session::new, "overwrite_no_truncate_test", write_only);
+}
+
+#[test_case(true; "write_only")]
+#[test_case(false; "read_write")]
+fn overwrite_no_truncate_test_mock(write_only: bool) {
+    overwrite_no_truncate_test(fuse::mock_session::new, "overwrite_no_truncate_test", write_only);
+}
+
+fn overwrite_truncate_test<F>(creator_fn: F, prefix: &str, write_only: bool)
```

Comment:
Does this test still have the race condition of file handle being opened for reading too soon after a write, resulting in bad FD?

Reply:
This is currently passing on `main`, we'll review again when changing behavior.

Reply:
Yes, you are right. But nothing has changed in this PR.